clients/dns-service-client/src/diff.rs

```diff
@@ -2,173 +2,249 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-use crate::types::DnsConfigParams;
+use crate::types::DnsConfigZone;
 use crate::types::DnsRecord;
+use crate::types::Srv;
 use crate::DnsRecords;
 use anyhow::ensure;
-use anyhow::Context;
+use std::collections::BTreeSet;
+
+#[derive(Debug)]
+enum NameDiff<'a> {
+    Added(&'a str, &'a [DnsRecord]),
+    Removed(&'a str, &'a [DnsRecord]),
+    Changed(&'a str, &'a [DnsRecord], &'a [DnsRecord]),
+    Unchanged(&'a str, &'a [DnsRecord]),
+}
 
 /// Compare the DNS records contained in two sets of DNS configuration
 #[derive(Debug)]
 pub struct DnsDiff<'a> {
     left: &'a DnsRecords,
     right: &'a DnsRecords,
+    zone_name: &'a str,
 }
 
 impl<'a> DnsDiff<'a> {
-    /// Compare the DNS records contained in two sets of DNS configuration
+    /// Compare the DNS records contained in two DNS zones' configs
     ///
-    /// Both configurations are expected to contain exactly one zone and they
-    /// should have the same name.
+    /// Both zones are expected to have the same name.
     pub fn new(
-        left: &'a DnsConfigParams,
-        right: &'a DnsConfigParams,
+        left_zone: &'a DnsConfigZone,
+        right_zone: &'a DnsConfigZone,
     ) -> Result<DnsDiff<'a>, anyhow::Error> {
-        let left_zone = left.sole_zone().context("left side of diff")?;
-        let right_zone = right.sole_zone().context("right side of diff")?;
-
         ensure!(
             left_zone.zone_name == right_zone.zone_name,
             "cannot compare DNS configuration from zones with different names: \
             {:?} vs. {:?}", left_zone.zone_name, right_zone.zone_name,
         );
 
-        Ok(DnsDiff { left: &left_zone.records, right: &right_zone.records })
+        Ok(DnsDiff {
+            left: &left_zone.records,
+            right: &right_zone.records,
+            zone_name: &left_zone.zone_name,
+        })
+    }
+
+    fn iter_names(&self) -> impl Iterator<Item = NameDiff<'_>> {
+        let all_names: BTreeSet<_> =
+            self.left.keys().chain(self.right.keys()).collect();
```

Comment:
Feel free to ignore this as premature optimization, but - should we cache `all_names`? If a caller calls multiple methods that all internally call `iter_names()`, we'll recreate this set every time.